File: src/web/static/js/key-statistics.js
/**
 * Key Statistics Module
 * Handles loading and displaying fundamental metrics
 */

class KeyStatisticsManager {
    constructor() {
        this.currentSymbol = null;
        this.statsData = null;
    }

    /**
     * Load key statistics for a symbol
     */
    async loadKeyStatistics(symbol) {
        if (!symbol) {
            console.error('No symbol provided');
            return;
        }

        this.currentSymbol = symbol;
        
        // Show loading indicator
        document.getElementById('fundamentalsLoading').style.display = 'block';
        document.getElementById('fundamentalsContent').style.display = 'none';

        try {
            const response = await fetch(`/api/key-statistics/${symbol}`);
            
            if (!response.ok) {
                if (response.status === 404) {
                    this.showNoDataMessage(symbol);
                    return;
                }
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();
            
            if (data.success) {
                this.statsData = data;
                this.displayKeyStatistics(data);
            } else {
                this.showErrorMessage('Failed to load key statistics');
            }

        } catch (error) {
            console.error('Error loading key statistics:', error);
            this.showErrorMessage(`Error: ${error.message}`);
        }
    }

    /**
     * Display key statistics in the UI
     */
    async displayKeyStatistics(data) {
        // Hide loading, show content
        document.getElementById('fundamentalsLoading').style.display = 'none';
        document.getElementById('fundamentalsContent').style.display = 'block';

        // Fetch company name from company info API
        try {
            const companyResponse = await fetch(`/api/company-info/${data.symbol}`);
            if (companyResponse.ok) {
                const companyData = await companyResponse.json();
                document.getElementById('fundCompanyName').textContent = companyData.name || data.symbol;
            } else {
                document.getElementById('fundCompanyName').textContent = data.symbol;
            }
        } catch (error) {
            document.getElementById('fundCompanyName').textContent = data.symbol;
        }

        // Header badges
        document.getElementById('fundSymbol').textContent = data.symbol;
        document.getElementById('fundDate').textContent = `As of: ${data.date}`;
        document.getElementById('fundSource').textContent = data.data_source.toUpperCase();

        const stats = data.data;

        // Large Market Cap Card
        this.setValue('fundMarketCapLarge', stats.valuation.market_cap_display);
        
        // Top Summary Cards
        this.setValue('fundPECard', this.formatNumber(stats.valuation.trailing_pe, 2));
        this.setValue('fundROECard', stats.profitability.roe_display);
        this.setValue('fundProfitMarginCard', stats.profitability.profit_margin_display);

        // Valuation Metrics
        this.setValue('fundMarketCap', stats.valuation.market_cap_display);
        this.setFormattedValue('fundEnterpriseValue', stats.valuation.enterprise_value);
        this.setValue('fundTrailingPE', this.formatNumber(stats.valuation.trailing_pe, 2));
        this.setValue('fundForwardPE', this.formatNumber(stats.valuation.forward_pe, 2));
        this.setValue('fundPEG', this.formatNumber(stats.valuation.peg_ratio, 2));
        this.setValue('fundPriceBook', this.formatNumber(stats.valuation.price_to_book, 2));
        this.setValue('fundPriceSales', this.formatNumber(stats.valuation.price_to_sales, 2));
        this.setValue('fundEVRevenue', this.formatNumber(stats.valuation.enterprise_to_revenue, 2));
        this.setValue('fundEVEBITDA', this.formatNumber(stats.valuation.enterprise_to_ebitda, 2));

        // Profitability Metrics
        this.setValue('fundProfitMargin', stats.profitability.profit_margin_display);
        this.setPercentage('fundOperatingMargin', stats.profitability.operating_margin);
        this.setPercentage('fundGrossMargin', stats.profitability.gross_margin);
        this.setPercentage('fundEBITDAMargin', stats.profitability.ebitda_margin);
        this.setValue('fundROE', stats.profitability.roe_display);
        this.setPercentage('fundROA', stats.profitability.return_on_assets);

        // Financial Health
        this.setFormattedValue('fundTotalCash', stats.financial_health.total_cash);
        this.setFormattedValue('fundTotalDebt', stats.financial_health.total_debt);
        this.setValue('fundDebtEquity', stats.financial_health.debt_to_equity_display);
        this.setValue('fundCurrentRatio', this.formatNumber(stats.financial_health.current_ratio, 2));
        this.setValue('fundQuickRatio', this.formatNumber(stats.financial_health.quick_ratio, 2));
        this.setFormattedValue('fundFCF', stats.financial_health.free_cash_flow);
        this.setFormattedValue('fundRevenue', stats.financial_health.revenue);
        this.setValue('fundEPS', stats.financial_health.earnings_per_share ? `$${stats.financial_health.earnings_per_share.toFixed(2)}` : 'N/A');

        // Growth Metrics
        this.setPercentage('fundRevenueGrowth', stats.growth.revenue_growth);
        this.setPercentage('fundEarningsGrowth', stats.growth.earnings_growth);

        // Trading Metrics
        this.setValue('fundBeta', this.formatNumber(stats.trading.beta, 2));
        this.setValue('fund52WHigh', stats.trading.fifty_two_week_high ? `$${stats.trading.fifty_two_week_high.toFixed(2)}` : 'N/A');
        this.setValue('fund52WLow', stats.trading.fifty_two_week_low ? `$${stats.trading.fifty_two_week_low.toFixed(2)}` : 'N/A');
        this.setValue('fund50DAvg', stats.trading.fifty_day_average ? `$${stats.trading.fifty_day_average.toFixed(2)}` : 'N/A');
        this.setValue('fund200DAvg', stats.trading.two_hundred_day_average ? `$${stats.trading.two_hundred_day_average.toFixed(2)}` : 'N/A');
        this.setFormattedVolume('fundAvgVolume', stats.trading.average_volume);

        // Dividends & Ownership
        this.setValue('fundDivYield', stats.dividends.dividend_yield_display);
        this.setValue('fundDivRate', stats.dividends.dividend_rate ? `$${stats.dividends.dividend_rate.toFixed(2)}` : 'N/A');
        this.setPercentage('fundPayoutRatio', stats.dividends.payout_ratio);
        this.setPercentage('fundInsiders', stats.shares.held_percent_insiders);
        this.setPercentage('fundInstitutional', stats.shares.held_percent_institutions);
        this.setFormattedVolume('fundShortInterest', stats.shares.shares_short);
    }

    /**
     * Set element value with fallback
     */
    setValue(elementId, value) {
        const element = document.getElementById(elementId);
        if (element) {
            element.textContent = value || 'N/A';
        }
    }

    /**
     * Format percentage value
     */
    setPercentage(elementId, value) {
        const element = document.getElementById(elementId);
        if (element) {
            if (value !== null && value !== undefined) {
                element.textContent = `${(value * 100).toFixed(2)}%`;
            } else {
                element.textContent = 'N/A';
            }
        }
    }

    /**
     * Format large numbers with K/M/B/T suffixes
     */
    setFormattedValue(elementId, value) {
        const element = document.getElementById(elementId);
        if (element) {
            if (value !== null && value !== undefined) {
                element.textContent = this.formatLargeNumber(value);
            } else {
                element.textContent = 'N/A';
            }
        }
    }

    /**
     * Format volume with K/M/B suffixes
     */
    setFormattedVolume(elementId, value) {
        const element = document.getElementById(elementId);
        if (element) {
            if (value !== null && value !== undefined) {
                element.textContent = this.formatVolume(value);
            } else {
                element.textContent = 'N/A';
            }
        }
    }

    /**
     * Format number with specified decimals
     */
    formatNumber(value, decimals = 2) {
        if (value === null || value === undefined) {
            return 'N/A';
        }
        return value.toFixed(decimals);
    }

    /**
     * Format large numbers with suffixes
     */
    formatLargeNumber(value) {
        if (value === null || value === undefined) {
            return 'N/A';
        }

        const absValue = Math.abs(value);
        const sign = value < 0 ? '-' : '';

        if (absValue >= 1e12) {
            return `${sign}$${(absValue / 1e12).toFixed(2)}T`;
        } else if (absValue >= 1e9) {
            return `${sign}$${(absValue / 1e9).toFixed(2)}B`;
        } else if (absValue >= 1e6) {
            return `${sign}$${(absValue / 1e6).toFixed(2)}M`;
        } else if (absValue >= 1e3) {
            return `${sign}$${(absValue / 1e3).toFixed(2)}K`;
        } else {
            return `${sign}$${absValue.toFixed(2)}`;
        }
    }

    /**
     * Format volume numbers
     */
    formatVolume(value) {
        if (value === null || value === undefined) {
            return 'N/A';
        }

        if (value >= 1e9) {
            return `${(value / 1e9).toFixed(2)}B`;
        } else if (value >= 1e6) {
            return `${(value / 1e6).toFixed(2)}M`;
        } else if (value >= 1e3) {
            return `${(value / 1e3).toFixed(2)}K`;
        } else {
            return value.toLocaleString();
        }
    }

    /**
     * Show message when no data is available
     */
    showNoDataMessage(symbol) {
        document.getElementById('fundamentalsLoading').style.display = 'none';
        document.getElementById('fundamentalsContent').style.display = 'block';
        
        document.getElementById('fundCompanyName').textContent = symbol;
        document.getElementById('fundSymbol').textContent = symbol;
        document.getElementById('fundDate').textContent = 'No data available';
        document.getElementById('fundSource').textContent = 'Load data first';
    }

    /**
     * Show error message
     */
    showErrorMessage(message) {
        document.getElementById('fundamentalsLoading').innerHTML = `
            <i class="fas fa-exclamation-triangle" style="color: #dc2626;"></i>
            <p style="color: #dc2626;">${message}</p>
        `;
    }
}

// Initialize global instance
window.keyStatisticsManager = new KeyStatisticsManager();

// Hook into symbol change events
document.addEventListener('DOMContentLoaded', function() {
    // Listen for symbol selection changes
    const symbolSelect = document.getElementById('symbolSelect');
    if (symbolSelect) {
        symbolSelect.addEventListener('change', function() {
            const symbol = this.value;
            if (symbol && window.keyStatisticsManager) {
                // Load key statistics when fundamentals tab is active
                const fundamentalsTab = document.querySelector('[data-tab="fundamentals"]');
                if (fundamentalsTab && fundamentalsTab.classList.contains('active')) {
                    window.keyStatisticsManager.loadKeyStatistics(symbol);
                }
            }
        });
    }

    // Listen for tab clicks
    const tabButtons = document.querySelectorAll('.tab-button');
    tabButtons.forEach(button => {
        button.addEventListener('click', function() {
            const tabName = this.getAttribute('data-tab');
            
            // Load key statistics when fundamentals tab is clicked
            if (tabName === 'fundamentals') {
                const symbolSelect = document.getElementById('symbolSelect');
                const symbol = symbolSelect ? symbolSelect.value : null;
                
                if (symbol && window.keyStatisticsManager) {
                    window.keyStatisticsManager.loadKeyStatistics(symbol);
                }
            }
        });
    });
});

File: src/web/api/key_statistics.py
"""
API routes for key statistics data
"""

from datetime import date
from typing import Optional

from fastapi import APIRouter, HTTPException, Query
from sqlalchemy import select

from src.shared.database.base import db_transaction
from src.shared.database.models.key_statistics import KeyStatistics

router = APIRouter(prefix="/api/key-statistics", tags=["key-statistics"])


@router.get("/{symbol}")
async def get_key_statistics(
    symbol: str,
    stats_date: Optional[str] = Query(None, description="Statistics date (YYYY-MM-DD)"),
) -> dict:
    """
    Get key statistics for a symbol

    Args:
        symbol: Stock symbol
        stats_date: Optional date filter (defaults to latest)

    Returns:
        Key statistics data
    """
    symbol = symbol.upper()

    try:
        with db_transaction() as session:
            query = select(KeyStatistics).where(KeyStatistics.symbol == symbol)

            if stats_date and isinstance(stats_date, str):
                query = query.where(
                    KeyStatistics.date == date.fromisoformat(stats_date)
                )

            query = query.order_by(KeyStatistics.date.desc())

            result = session.execute(query).first()

            if not result:
                raise HTTPException(
                    status_code=404, detail=f"No key statistics found for {symbol}"
                )

            stats = result[0]

            return {
                "success": True,
                "symbol": stats.symbol,
                "date": stats.date.isoformat(),
                "data_source": stats.data_source,
                "data": {
                    # Valuation metrics
                    "valuation": {
                        "market_cap": stats.market_cap,
                        "market_cap_display": stats.market_cap_display,
                        "enterprise_value": stats.enterprise_value,
                        "trailing_pe": (
                            float(stats.trailing_pe) if stats.trailing_pe else None
                        ),
                        "forward_pe": (
                            float(stats.forward_pe) if stats.forward_pe else None
                        ),
                        "peg_ratio": (
                            float(stats.peg_ratio) if stats.peg_ratio else None
                        ),
                        "price_to_book": (
                            float(stats.price_to_book) if stats.price_to_book else None
                        ),
                        "price_to_sales": (
                            float(stats.price_to_sales)
                            if stats.price_to_sales
                            else None
                        ),
                        "enterprise_to_revenue": (
                            float(stats.enterprise_to_revenue)
                            if stats.enterprise_to_revenue
                            else None
                        ),
                        "enterprise_to_ebitda": (
                            float(stats.enterprise_to_ebitda)
                            if stats.enterprise_to_ebitda
                            else None
                        ),
                    },
                    # Profitability metrics
                    "profitability": {
                        "profit_margin": (
                            float(stats.profit_margin) if stats.profit_margin else None
                        ),
                        "profit_margin_display": stats.profit_margin_display,
                        "operating_margin": (
                            float(stats.operating_margin)
                            if stats.operating_margin
                            else None
                        ),
                        "return_on_assets": (
                            float(stats.return_on_assets)
                            if stats.return_on_assets
                            else None
                        ),
                        "return_on_equity": (
                            float(stats.return_on_equity)
                            if stats.return_on_equity
                            else None
                        ),
                        "roe_display": stats.roe_display,
                        "gross_margin": (
                            float(stats.gross_margin) if stats.gross_margin else None
                        ),
                        "ebitda_margin": (
                            float(stats.ebitda_margin) if stats.ebitda_margin else None
                        ),
                    },
                    # Financial health
                    "financial_health": {
                        "revenue": stats.revenue,
                        "revenue_per_share": (
                            float(stats.revenue_per_share)
                            if stats.revenue_per_share
                            else None
                        ),
                        "earnings_per_share": (
                            float(stats.earnings_per_share)
                            if stats.earnings_per_share
                            else None
                        ),
                        "total_cash": stats.total_cash,
                        "total_debt": stats.total_debt,
                        "debt_to_equity": (
                            float(stats.debt_to_equity)
                            if stats.debt_to_equity
                            else None
                        ),
                        "debt_to_equity_display": stats.debt_to_equity_display,
                        "current_ratio": (
                            float(stats.current_ratio) if stats.current_ratio else None
                        ),
                        "quick_ratio": (
                            float(stats.quick_ratio) if stats.quick_ratio else None
                        ),
                        "free_cash_flow": stats.free_cash_flow,
                        "operating_cash_flow": stats.operating_cash_flow,
                    },
                    # Growth metrics
                    "growth": {
                        "revenue_growth": (
                            float(stats.revenue_growth)
                            if stats.revenue_growth
                            else None
                        ),
                        "earnings_growth": (
                            float(stats.earnings_growth)
                            if stats.earnings_growth
                            else None
                        ),
                    },
                    # Trading metrics
                    "trading": {
                        "beta": float(stats.beta) if stats.beta else None,
                        "fifty_two_week_high": (
                            float(stats.fifty_two_week_high)
                            if stats.fifty_two_week_high
                            else None
                        ),
                        "fifty_two_week_low": (
                            float(stats.fifty_two_week_low)
                            if stats.fifty_two_week_low
                            else None
                        ),
                        "fifty_day_average": (
                            float(stats.fifty_day_average)
                            if stats.fifty_day_average
                            else None
                        ),
                        "two_hundred_day_average": (
                            float(stats.two_hundred_day_average)
                            if stats.two_hundred_day_average
                            else None
                        ),
                        "average_volume": stats.average_volume,
                    },
                    # Dividend metrics
                    "dividends": {
                        "dividend_yield": (
                            float(stats.dividend_yield)
                            if stats.dividend_yield
                            else None
                        ),
                        "dividend_yield_display": stats.dividend_yield_display,
                        "dividend_rate": (
                            float(stats.dividend_rate) if stats.dividend_rate else None
                        ),
                        "payout_ratio": (
                            float(stats.payout_ratio) if stats.payout_ratio else None
                        ),
                    },
                    # Share information
                    "shares": {
                        "shares_outstanding": stats.shares_outstanding,
                        "float_shares": stats.float_shares,
                        "shares_short": stats.shares_short,
                        "short_ratio": (
                            float(stats.short_ratio) if stats.short_ratio else None
                        ),
                        "held_percent_insiders": (
                            float(stats.held_percent_insiders)
                            if stats.held_percent_insiders
                            else None
                        ),
                        "held_percent_institutions": (
                            float(stats.held_percent_institutions)
                            if stats.held_percent_institutions
                            else None
                        ),
                    },
                },
                "updated_at": (
                    stats.updated_at.isoformat() if stats.updated_at else None
                ),
            }

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to fetch key statistics: {str(e)}"
        )


@router.get("")
async def list_available_symbols() -> dict:
    """
    Get list of symbols that have key statistics data

    Returns:
        List of symbols with their latest statistics date
    """
    try:
        with db_transaction() as session:
            query = (
                select(KeyStatistics.symbol, KeyStatistics.date)
                .order_by(KeyStatistics.symbol, KeyStatistics.date.desc())
                .distinct(KeyStatistics.symbol)
            )

            results = session.execute(query).all()

            symbols = [
                {"symbol": row[0], "latest_date": row[1].isoformat()} for row in results
            ]

            return {"success": True, "count": len(symbols), "symbols": symbols}

    except Exception as e:
        raise HTTPException(
            status_code=500, detail=f"Failed to fetch symbols: {str(e)}"
        )
